course7/homeworks/mart/complex/complex.cc

```diff
@@ -0,0 +1,155 @@
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <stack>
+#include <regex>
+
+struct Complex
+{
+	Complex() = default;
+	Complex(double r) : real(r) {}
+	Complex(double r, double i) : real(r) , imaginary(i) {}
+	Complex(const Complex& src) : real(src.real) , imaginary(src.imaginary) {}
+
+	Complex& operator=(const Complex& src)
+	{
+		real = src.real;
+		imaginary = src.imaginary;
+		return *this;
+	}
+
+	Complex operator+(const Complex& rhs) const
+	{
+		return { real + rhs.real, imaginary + rhs.imaginary };
+	}
+
+	Complex operator-(const Complex& rhs) const
+	{
+		return { real - rhs.real, imaginary - rhs.imaginary };
+	}
+
+	Complex operator*(const Complex& rhs) const
+	{
+		return {
+			(real * rhs.real) - (imaginary * rhs.imaginary),
+			(real * rhs.imaginary) + (imaginary * rhs.real)
+		  };
+	}
+
+	Complex operator/(const Complex& rhs) const
+	{
+		return {
+			((real * rhs.real) + (imaginary * rhs.imaginary)) / ((rhs.real * rhs.real) + (rhs.imaginary * rhs.imaginary)),
+			((imaginary * rhs.real) - (real * rhs.imaginary)) / ((rhs.real * rhs.real) + (rhs.imaginary * rhs.imaginary))
+		  };
+	}
+
+	bool operator==(const Complex& rhs)
+	{
+		return real == rhs.real && imaginary == rhs.imaginary;
+	}
+
+	double real = 0.0;
+	double imaginary = 0.0;
+};
+
+std::ostream& operator<<(std::ostream& stream, const Complex& complex)
+{
+	stream << complex.real << "+i" << complex.imaginary;
+	return stream;
+}
+
+class Calculator
+{
+public:
+	void run()
+	{
+		for (std::string line; std::getline(std::cin, line); )
+		{
+			try
+			{
+				if (line == "+")
+					add();
+				else if (line == "-")
+					subtract();
+				else if (line == "*")
+					multiply();
+				else if (line == "/")
+					divide();
+				else
+					parseAndPushComplex(line);
+			}
+			catch(const std::exception& e)
+			{
+				std::cerr << "Error executing line: " << e.what() << std::endl;
```

Comment:
I think that after an error the calculator would be in an inconsistent state, so the error should be fatal.
Therefore, it would be better to let the exception fly trough.